File: Java/20-ValidParentheses.java
public class Solution {
    public boolean isValid(String s) {
        if (s.length() == 0)
            return true;
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char curr = s.charAt(i);
            if ((curr == '}' || curr == ']' || curr == ')') && stack.isEmpty()) 
                return false;
            if (curr == '}' && stack.peek() != '{')
                return false;
            if (curr == ']' && stack.peek() != '[')
                return false;
            if (curr == ')' && stack.peek() != '(')
                return false;
            if (curr == '{' || curr == '(' || curr == '[') {
                stack.push(curr);
            } else {
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
}

//another solution
public class Solution {
    public boolean isValid(String s) {
        if (s == null || s.length() == 0){
            return true;
        }
        List<Character> left = Arrays.asList('(', '{', '[');
        List<Character> right = Arrays.asList(')', '}', ']');
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++){
            char curr = s.charAt(i);
            if (left.contains(curr)){
                stack.push(curr);
            }else if (!stack.isEmpty()){
                if (stack.peek() == left.get(right.indexOf(curr))){
                    stack.pop();
                    continue;
                }else{
                    return false;
                }
            }else{
                return false;
            }
        }
        return stack.isEmpty();
    }
}

File: cpp/20-ValidParentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_set<char> left = {'(', '{', '['};
        unordered_set<char> right = {')', '}', ']'};
        for (auto c : s){
            if (left.count(c)){
                stack.push(c);
            }else{
                if (stack.empty()){
                    return false;
                }
                if (c != ')' && stack.top() == '('){
                    return false;
                }
                if (c != '}' && stack.top() == '{'){
                    return false;
                }
                if (c != ']' && stack.top() == '[]'){
                    return false;
                }
                stack.pop();
            }
        }
        return stack.empty();
    }
};